SaraAlert/SaraAlert
# SARAALERT-1539: Common exposure cohort fields

# Description

Jira Ticket: [SARAALERT-1539](https://tracker.codev.mitre.org/browse/SARAALERT-1539)

See JIRA ticket for description and [mockups](https://teams.microsoft.com/l/file/033C5559-68D8-4C59-9833-3F020415D877?tenantId=c620dc48-1d50-4952-8b39-df4d54d74d82&fileType=pdf&objectUrl=https%3A%2F%2Fmitre.sharepoint.com%2Fsites%2FCoVDT%2FShared%20Documents%2FGeneral%2FProduct%20Owner%20Materials%2FRequirements%2FMockups%2FExposure%20Cohort%20(1539)%20Jul%2012%202021.pdf&baseUrl=https%3A%2F%2Fmitre.sharepoint.com%2Fsites%2FCoVDT&serviceName=teams&threadId=19:00d3d45c64ca457795d58c2764d517cd@thread.skype&groupId=7de0f318-3170-4f1c-82eb-95bfe132bfe9)

## (Feature) Demo/Screenshots

![Screen Shot 2021-10-27 at 3 52 02 PM](https://user-images.githubusercontent.com/9956561/139137176-37f2eb52-3344-4fd6-b88f-87bb9c8c2cf7.png)
![Screen Shot 2021-10-27 at 3 52 22 PM](https://user-images.githubusercontent.com/9956561/139137180-5dd19082-36a2-4b31-a474-00e1b6a98eb0.png)
![Screen Shot 2021-11-10 at 2 26 16 AM](https://user-images.githubusercontent.com/9956561/141068643-da13bc5e-dc53-4c6b-9e17-0f56bb911e2c.png)
![Screen Shot 2021-11-10 at 2 27 16 AM](https://user-images.githubusercontent.com/9956561/141068648-2e776411-aead-4e43-b673-2b38fb7e8000.png)

## Important Changes

`app/controllers/jurisdictions_controller.rb`
- new controller methods to query existing cohort names and locations

`app/controllers/patients_controller.rb`
- support creating cohorts for new patients and creating/updating/deleting cohorts for existing patients

`app/helpers/patient_helper.rb`
- offload allowed_params from patients_controller to here because of rubocop

`app/javascript/components/enrollment/steps/ExposureInformation.js`
- add cohort components

`app/javascript/components/patient/Patient.js`
- add cohorts to monitoree details page

`app/models/concerns/associated_record_limit_validator.rb`
- limit patients to 10 cohorts max

`app/jobs/cache_analytics_job.rb`
- update cohort query in analytics job

`app/lib/import_export_constants.rb`
- support importing/exporting cohorts

`db/migrate/20211027041647_create_common_exposure_cohorts.rb`
- migrate common exposure cohort data

`lib/tasks/demo.rake`
- refactor demo script, populate cohorts

`public/Sara Alert Import Format.xlsx`
- update formatting guidance

## Testing Recommendations

- Migration/Rollback
  - Should create/remove table, add/remove columns
  - Should create cohorts based on existing cohorts
  - Should update advanced filters in both `user_filters` and `user_export_presets`
  - Should update user export preset if `common_exposure_cohort_type` is selected
  - Should not modify patient timestamps
  - Should rollback all operations if failure is present, error printed, migration marked as failed
- Enrollment Wizard/Monitoree Details Page
  - Add/Edit/Delete cohorts for new and existing patients
  - Verify data lists
- Advanced Filter
  - Filter by multiple cohort fields
- Import/Export
  - Ensure cohorts can be properly imported and exported
- API
  - Create patient via patient endpoint and transaction endpoint
- Demo Script
  - Ensure common exposure cohort data is populated

# Checklists

**Submitter:**


@tstrass :


@ngfreiter :


@sgober :
